### ISSUE TITLE
Format esgf-world catalog to help with the netcdf->zarr pipeline

### ISSUE THREAD

**Issue by aradhakrishnanGFDL** [pangeo-forge/cmip6-pipeline#17]
This issue tracks the action needed to keep the CSV header for the [esgf-world intake esm catalog ](https://cmip6-nc.s3.us-east-2.amazonaws.com/esgf-world.csv.gz) similar to that of the [zarr catalog](https://cmip6-pds.s3.amazonaws.com/pangeo-cmip6-noQC.csv.gz) to enable easy comparison and integration into the CMIP6 pipeline.

**Comment by rabernat** [comment]
For reference, we tried to follow the official [controlled vocabulary](https://github.com/WCRP-CMIP/CMIP6_CVs) as close as possible.

**Comment by aradhakrishnanGFDL** [comment]
@naomi-henderson @rabernat 
Our student intern is running into issues with some of the older versions of datasets that are flawed -- and I am considering one of the following changes to the esgf-world intake catalog. Any thoughts on what's better or what you recommend based on how the zarr intake catalogs are populated?

1) ensure latest versions are the only ones that show up in the catalog because some of the old versions are flawed 
2) set latest_version=True to the esgf-world intake 

 Having a single version in the esgf-world catalog (or having an easier way to identify the latest) should also help with reducing some of the data prep steps in the pangeo-forge recipe (cmip6 zarr to netcdf pipeline)

**Comment by naomi-henderson** [comment]
@aradhakrishnanGFDL , thanks for opening this issue!  Of course your s3://esgf-world pathnames give the CV values directly:
```
s3://esgf-world/CMIP6/activity_id/institution_id/source_id/experiment_id/member_id/table_id/variable_id/grid_label/version/*.nc
```
The netcdf file name itself could still be parsed for what you call the 'temporal subset'. Note that extra columns in the csv files do not cause trouble, but probably best if you keep them further to the right, for readability.

I also suggest that you keep just the one master csv.gz file which will be read by `intake-esm`.  It will be simpler in the long run, I regret our multiple names for the same file since we cannot make links in the object store!   But this means that you will need to modify your `intake-esm` JSON file accordingly.

**Comment by naomi-henderson** [comment]
@aradhakrishnanGFDL , yes the multiple versions cause trouble, but I just spent many weeks last fall making it so that we can have multiple versions!!  There are a few problems that come up.  When you update the bucket with new versions, will you remember to delete the old?  If you delete the old, will it break old, working code?  I was getting many more questions about why the science questions were getting different answers in the past than now. Often there was bad data in the old versions but there was no way to find out which version and what the problem was!  

Instead, I suggest you encourage everyone to do a quick check and grab only the latest version.  Then we don't pull the rug out from anyone.

**Comment by naomi-henderson** [comment]
Doesn't `intake-esm` only use the latest version by default?  Somehow I thought it did.

**Comment by aradhakrishnanGFDL** [comment]
Hi @naomi-henderson 

I've updated the esgf-world intake catalogs to reflect the same header names as the zarr catalog, except for "path" (vs zstore). 

https://github.com/aradhakrishnanGFDL/gfdl-aws-analysis/blob/master/esm-collection-spec-examples/esgf-world.json

The changes to the header are only for those columns that existed previously in the esgf-world catalog. I have not attempted to add any new columns,e.g there is no activity_id in the above version. 

Let me know if the catalog in its current form helps a bit and if there are any other suggestions on this.

zarr->nc pangeo forge recipe.

Re: versioning, in the tests with Mackenzie Blanusa,  when the latest versions differ across all (netcdf) datasets in the search, I do not believe the latest version was being returned for each one of the returned xarray dataset objects. But, this could be a separate issue/topic for discussion later. For now, @jbusecke provided a pointer or two and I am yet to look into that. I'll open another issue to track the versioning stuff if needed.

**Comment by naomi-henderson** [comment]
@aradhakrishnanGFDL, this version of the `JSON` and `esgf-world.csv` should work much better.  Thanks

Re: versioning, this is a bit vague.  What are you using to find the latest versions? What tests did Mackenzie do and which code is returning a version which might not be the latest?  Are you still referring to `intake-esm`?  If so, I can't really help with this since I don't use `intake-esm` - perhaps @andersy005 could help with this?

**Comment by aradhakrishnanGFDL** [comment]
Hi @naomi-henderson I think the versioning issue should be a separate github issue. I have example notebooks to explain this further and will open another issue if necessary. No action needed on "versioning" at this time. Thank you. 

> @aradhakrishnanGFDL, this version of the `JSON` and `esgf-world.csv` should work much better. Thanks
> 
> Re: versioning, this is a bit vague. What are you using to find the latest versions? What tests did Mackenzie do and which code is returning a version which might not be the latest? Are you still referring to `intake-esm`? If so, I can't really help with this since I don't use `intake-esm` - perhaps @andersy005 could help with this?

**Comment by andersy005** [comment]
@aradhakrishnanGFDL, 

To avoid dealing with complications caused by the presence of multiple versions, we decided to do what proposed in your comment above:

> ensure latest versions are the only ones that show up in the catalog because some of the old versions are flawed


Here's a function I use to discard old versions when building our CMIP6 archive catalog:


```python
import itertools
def pick_latest_version(df):

    print(f'Dataframe size before picking latest version: {len(df)}')
    grpby = list(set(df.columns.tolist()) - {'path', 'version'})
    grouped = df.groupby(grpby)

    def _pick_latest_v(group):
        idx = []
        if group.version.nunique() > 1:
            idx = group.sort_values(by=['version'], ascending=False).index[1:].values.tolist()
        return idx

    print('Getting latest version...\n')
    idx_to_remove = grouped.apply(_pick_latest_v).tolist()
    idx_to_remove = list(itertools.chain(*idx_to_remove))
    df = df.drop(index=idx_to_remove)
    print(f'Dataframe size after picking latest version: {len(df)}')
    print('\nDone....\n')
    return df
```

I called this function on your catalog and things seem to be working:

```python
In [11]: col.df.shape
Out[11]: (1066084, 13)

In [12]: df = pick_latest_version(col.df)
Dataframe size before picking latest version: 1066084
Getting latest version...

Dataframe size after picking latest version: 996342

Done....


In [13]: df.shape
Out[13]: (996342, 13)
```
